WIP: Plugin API redesign v2.1

algernon opened the pull request:
This is a work in progress, building on top of #276.

noseglasses commented on doc/plugin-api-internals.md:
````diff
@@ -0,0 +1,488 @@
+Which in turn becomes:
+
+```c++
+static auto apply(Key &mappedKey, byte row, byte col, uint8_t keyState) {
````
```cpp
static auto apply(Key &mappedKey, byte row, byte col, uint8_t keyState) {
```
should be 
```cpp
static bool apply(Key &mappedKey, byte row, byte col, uint8_t keyState) {
```
to be valid C++11

noseglasses commented on doc/plugin-api-internals.md:
````diff
@@ -0,0 +1,488 @@
+
+```c++
+static auto apply(Key &mappedKey, byte row, byte col, uint8_t keyState) {
+  ContinueIfHookReturnsTrue hook_return_val;
````
`ContinueIfHookReturnsTrue` is actually a class, better replace it with `bool` here.

noseglasses replied:
Most precise would be 
```cpp
decltype(ContinueIfHookReturnsTrue::eval(true)) hook_return_val;
```
but that's of no use here.

noseglasses commented on doc/plugin-api-internals.md:
````diff
@@ -0,0 +1,488 @@
+we end up with:
+
+```c++
+static auto apply(Key &mappedKey, byte row, byte col, uint8_t keyState) {
````
Same here
```cpp
static auto apply(Key &mappedKey, byte row, byte col, uint8_t keyState) {
``

algernon commented:
We talked with @obra at length last night, about naming things, and came up with the following:

- `OrderedPlugins` would become `HookPoint`, because it offers a way to call a collection of plugin methods for a particular hook point.
- `HookTask_*` would become `pluginMethod_*`, because it wraps that.
- We'd invert `ContinuationPredicate` (because continue is the default), and rename it to `ShouldAbort`, adjusting the code as necessary. Inverting, because continue should be the default, rename because "predicate" is a great word, but `ShouldAbort` describes the same intent in easier to understand words.
- We'd also rename `HookTask_*::invoke` to `pluginMethod_*::call`, because we felt it sounded better.
- Hooks would also get new names, to clean up the mess we had before. They'll follow the same pattern, and better describe when they're called. We're also getting rid of "Hook" in their name.
  - `eventHandlerHook` becomes `onEvent`.
  - `preReportHook` becomes `beforeReportingState`
  - `postReportHook` becomes `afterReportingState`
  - `init` becomes `onSetup`
  - A separate PR will introduce `beforeEachCycle` (and/or `beforeGatheringEvents`), `afterGatheringEvents`, `afterEachCycle` (possibly replacing `afterReportingState`).

algernon commented:
It would be, and we went with `PluginMethod_` indeed. That was the original plan too, I just typoed previously O:)

algernon commented:
I think the code is now ready for review. I still have to clean up the history a little bit, but that shouldn't change the end result.

obra commented:
14:57 <@obra>  Looking at the amount of code we have for dealing with returntypes makes me really want to enforce the same return type on everything. I'm not saying that we Must do that, but it feels like we 
              should.
14:58 <@algernon> I wouldn't be opposed to that, if we make that return type an enum
14:58 <@algernon> that's ~extensible
14:58 <@algernon> also makes plugin code easier to understand
14:59 <@algernon> right now, you need to *know* that onEvent should return true to continue, false to abort
14:59 <@obra> The current values  would be SUCCESS, DID_NOT_RUN, FAILURE?
14:59 <@algernon> for someone reading onEvent, without prior knowledge of how onEvents are chained, will be rightly confused
15:00 <@obra> though those don't actually say whether we would want to stop processing
15:00 <@algernon> right now, only onEvent uses the return value, and there the values would be CONTINUE and ABORT
15:00 <@algernon> or perhaps NEXT/EXIT
15:01 <@algernon> return NEXT => let the next handler have a go at it; return EXIT => bye-bye
15:01 <@obra> since this isn't 1970, we can use more verbose names :)
15:01 <@algernon> oh, right. we had a power outage and the clock reset to the epoch!
15:01  * algernon adjusts
15:04 <@algernon> return NEXT_HANDLER (or NEXT_IN_CHAIN, or CONTINUE_WITH_NEXT_HANDLER) + return WITHOUT_FURTHER_PROCESSING
15:05 <@obra> we don't really call them handlers in this context.
15:06 <@obra> well, not after that renaming, anyway
15:08 <@obra> return ABORT_HANDLING ABORT_PROCESSING EVENT_CONSUMED
15:08 <@obra> return DONE_PROCESSING
15:08 <@obra> return CANCEL_PROCESSING_EVENT
15:09 <@algernon> I like EVENT_CONSUMED best so far
15:09 <@obra> sure. we could also spec that that enum value only makes sense in the onEvent handler
15:10 <@algernon> it says what the plugin considers, not what the mechanism outside of it will do
15:10 <@obra> right
15:10 <@obra> return OK for the other default case
15:10 <@algernon> that sounds good.
15:11 <@obra> it probably makes sense to have a return ERROR there too, even if we don't typically act on it.
15:11 <@obra> I could see a debugging build where we do something with that
15:11 <@algernon> btw, can you dump this in #316 too? (along with anything else you'd l

obra commented:
15:13 <@algernon> comment cleanup, inlining macros (_call_*, HOOK_STATIC_METHODS), "why is this in _internal?" (hook_rooting), moving stuff around (MAP out of internal, etc)
15:14 <@algernon> tbh, if you paste my line above, that should be enough, I'll find the logs from there =)

noseglasses commented on src/macro_helpers.h:
```diff
@@ -81,43 +81,43 @@ int array[] = { A, B, RESTRICT_ARGS_COUNT(C, 3, B_MACRO, ##__VA_ARGS__) };
-__NL__   )
+   (                                                                      __NL__ \
+   []{ /* Here we are in the body of a dummy lambda function.             __NL__ \
+          []{} is, BTW, the shortest way to write a lambda.               __NL__ \
```
The `__NL__` macro is meant to improve readability of pre-processed code. Adding it to lines that contain only comments will result in a bunch of empty lines in the resulting pre-processed code. I suppose that's unwanted. Because of this, the `__NN__` macro was added. Its sole purpose is not to break the flow of newline macros in the original code and to improve its readability without generating undesired newlines for these lines after pre-processing.

obra replied:
Yeah. My thought process was around the tradeoff between the generated code being more compact and the human-editable version of the code being more maintainable. 
Ultimately, I expect that folks will be spending a lot more time with the 'real' implementation than the generated code. And some extra newlines in the generated code feel to me like they're not a significant hardship relative to making editing the code more complicated.

I've actually been wondering if astyle would do a good enough job with the generated code that we could instead just get rid of the __NL__ macros entirely.

noseglasses replied:
> And some extra newlines in the generated code feel to me like they're not a significant hardship relative to making editing the code more complicated.

Agreed.

> I've actually been wondering if astyle would do a good enough job with the generated code that we could instead just get rid of the NL macros entirely.

As far as I get it, astyle is currently processing the original code. This won't fix the problems that arise with multi-line pre-processor macros. Every `\` followed by a `\n` is completely removed, and the whole macro ends up in a single, possibly very long line. That's why the `__NL__` is useful.

Of course you could run astyle on the pre-processed code, it will probably add some line breaks. But I prefer the pre-processed code to have the same layout as the original code which aids orientation and code comparison.

noseglasses commented:
Wow, I am impressed how fast you are progressing with this. May I add some remarks about the recent changes.

>  it probably makes sense to have a return ERROR there too, even if we don't typically act on it.

I like the idea of reducing the complexity by using a single return value type. However, your discussion suggests that you intent to add values that are possibly ignored. That sounds like a pretty bad idea, unless it is perfectly documented. As a programmer I always expect the calling instance to react reasonably on my return codes and I rely on that. When return codes are ignored, in general that means a design flaw or at least an obvious trap. That's why I would suggest to only define those values that are really required and acted upon. 
Passing error codes around in a firmware doesn't seem like a good idea either. You don't want to have release builds with error handling. Error handling code for debug versions is not a good idea either. I'd encourage to use assertions for that purpose. They tend to improve code readability as they describe what is expected. 

The Kaleidoscope coding style guide prefers [enums to be named as constants](https://github.com/keyboardio/Kaleidoscope/blob/master/doc/CODING_STYLE.md#enumerator-names). I would strongly recommend to use all-caps names __for macros only__. This makes it possible to rely on what is a symbol (no-strange side effects from replacement & type safety) and what is a macro (be prepared for any surprises). This differentiation, if taken serious, helps a lot when debugging.

Is there any reason why you intent to use plain old enums instead of enum classes for the handlers' return values? Unless there are really good reasons against, the latter should be preferred. See [here](https://stackoverflow.com/questions/18335861/why-is-enum-class-preferred-over-plain-enum) about the difference between both enum types.

I read that you don't intent to call the plugins' dedicated callback methods 'handlers'. How do you intent to call them instead? In other contexts like GUI frameworks this is exactly what would be called event/signal handlers. The only difference here is that they are not dynamically registered. Nevertheless, the name seems appropriate to me.

gedankenexperimenter commented:
> I would strongly recommend to use all-caps names for macros only.

Seconded.

> I read that you don't intent to call the plugins' dedicated callback methods 'handlers'. How do you intent to call them instead? In other contexts like GUI frameworks this is exactly what would be called event/signal handlers. The only difference here is that they are not dynamically registered.

In the case of the event handlers, the name "handler" is certainly appropriate, but many of the hook functions don't actually "handle" anything. And in many cases, even the "event handler" hook functions might do something to _respond_ to an event, but don't actually _handle_ that event completely, so one could argue that "handler" is a bit of a misnomer even there.

noseglasses commented:
You are right, handlers aren't forced to handle. There are at least three names of this concept I am aware of: events&handlers, signals&slots and hooks. Maybe there are others. In any case, I think we should stick with conventions and use one of the established terms.

noseglasses commented on src/kaleidoscope_internal/hook_rooting.h:
```diff
@@ -0,0 +1,257 @@
+                                                                          __NL__ \
+   struct PluginMethod_##PLUGIN_METHOD {                                  __NL__ \
+                                                                          __NL__ \
+      typedef SHOULD_ABORT_TYPE shouldAbort;                              __NL__ \
```
`shouldAbort` might better be named uppercase as it is a type not an instance.

algernon commented:
TODO list for myself:

- [ ] Figure out a better name for `hook_rooting.h`
- [ ] Explain at the top of said file why it is in `src/kaleidoscope_internal`
- [ ] Move `preprocessor_macro_map.h` elsewhere, or present a good case why it should be internal (even though it is just preprocessor macros, so not namespaced!)
- [ ] Figure out what to do with return types. Having them all be the same certainly has advantages.

obra commented:
@noseglasses:
> events&handlers, signals&slots and hooks.

In Kaleidoscope, we don't use the 'signals and slots' nomenclature for anything.
'hooks' are the places where plugins can provide implementations of functionality.
'events' are physical changes to key state (or things that pretend to be physical changes)
'handlers' are the functions in plugins that satisfy hooks.

noseglasses commented:
@obra:

>In Kaleidoscope, we don't use the 'signals and slots' nomenclature for anything.
'hooks' are the places where plugins can provide implementations of functionality.
'events' are physical changes to key state (or things that pretend to be physical changes)
'handlers' are the functions in plugins that satisfy hooks.

I was refering to this part of your communication:

> @algernon> return NEXT_HANDLER (or NEXT_IN_CHAIN, or CONTINUE_WITH_NEXT_HANDLER) + return WITHOUT_FURTHER_PROCESSING
15:05 <@obra> we don't really call them handlers in this context.
15:06 <@obra> well, not after that renaming, anyway

Which appears to be contradicted by your more recent post.

> 'handlers' are the functions in plugins that satisfy hooks.

FWIW, I spend enough time with this piece of code that I am well aware that there are no signals&slots :wink:

noseglasses commented:
> Yup, we're very much still feeling out the names.

Yes, and I am very happy with most of the names you found. I really appreciate you taking so much time to name everything appropriately. Correct and telling names will help future developers a lot.

> Right now, 'Handler' only actually gets used for key events.

I see. So what do we call the other callbacks then. From my perspective, any of these methods have a signal character. Forget about the slots here (I never liked that name anyway). In many other projects those methods would probably be called signal handlers. To me it sounds a bit strange 'to handle a signal', but I am not a native speaker. Anyway, it is quite commonly used that way.

The way you call the methods `on...`, `after...`, `before...` are also quite typical for the signal/handler concept. Those methods signal what just happened or what is going to happen next. Gtk/[Gtkmm](https://developer.gnome.org/gtkmm/stable/classGtk_1_1Widget.html#a81ec450f7c382fbeb54f7531b5ccde39) e.g. use a very similar naming.

algernon commented:
Just pushed a new update that changes the return type of all plugin methods to `kaleidoscope::Plugin::Result` (an enum class). We see a little bit of PROGMEM increase, but it does not appear to grow with the number of plugins enabled. I haven't checked if the compiler is smart enough to optimize out the shouldAbort checks at compile-time with this setup, but even if it doesn't, the minor code size increase, and the possible (but also very tiny) performance impact is well worth it for the simplified code.

Next up: cleaning up the history, and updating all the plugins.

noseglasses commented on src/plugin.h:
```diff
@@ -36,6 +36,11 @@ class Plugin {
 #endif
 
  public:
+  enum class Result {
```
As `enum class Result` is used by both the plugins' methods but also by the global hook functions, I am not sure if it belongs in class Plugin. Putting something in a class should either be because of the intention to use its namespace to associate or because of encapsulation. Both is not strictly the case here. The former because `Plugin::Result` sounds strange as it suggests that a plugin (not the handlers) has a result and the latter as the enum class must be public anyway.

I therefore suggest to move the enum class to `namespace kaleidoscope`. Although algernon and Obra seem not to have decided yet what to call the plugins' handler methods, calling the enum class `HandlerResult` or `HookResult` or `CallbackResult` seems more expressive.

noseglasses replied:
The members of `enum class Result` are still all caps. That's misleading as to most C++ programmers it will suggest that `OK`, `EVENT_CONSUMED` and `ERROR` are pre-processor macros.

algernon replied:
To me, `Plugin::Result` suggests that its the result of a plugin task. The reason it is under `Plugin` is because we wanted to tie it to plugins some way, without having to figure out what to call the handler methods :P

Regarding caps: I can be convinced to change that, as they are, indeed, not pre-processor macros.

noseglasses replied:
> without having to figure out what to call the handler methods :P

That's the first time that I am confronted with using classes to avoid proper naming :wink:
I will try to remember it for the next time that someone asks me about the reasons for using OOP.

noseglasses commented on src/kaleidoscope_internal/plugin_hooks.h:
```diff
@@ -90,118 +90,56 @@ namespace kaleidoscope_internal {
+       result == kaleidoscope::Plugin::Result::EVENT_CONSUMED) {     __NL__ \
       return result;                                                 __NL__ \
    }                                                                 __NL__
 
```
Currently the `Result` option `ERROR` is completely ignored and thus problematic as it might be returned intentionally. I would at least add an assertion to the end of `_CALL_HOOK_FOR_PLUGIN`
```cpp
assert(result != HandlerResult::Error);
```

Mind you that `-DNDEBUG` is probably still missing from the kaleidoscope_builder's default compile flags.

algernon replied:
It is ignored now on purpose: in case of an error, we want the rest of the chain to process it still. At least until we figure out a way to signal that error to the end-user.

Good point about assert & NDEBUG, though.

noseglasses replied:
I am afraid, I am still not convinced.

> until we figure out a way to signal that error to the end-user.

I suppose the resources of the firmware are too limited to even think about any kind of runtime error handling. There is no point in passing error flags around. It will always require proper debugging to fix those errors and that's where assertions come in handy. 

> It is ignored now on purpose: in case of an error, we want the rest of the chain to process it still.

Even if you stick with the `ERROR` return value, an assertion does ignore it in release builds. So it definitely does what you want. Only in debug builds the assertion will report it - until you come up with a better idea. Even though it is not yet possible to generate debug builds with kaleidsocope_builder, it definitely makes sense to add as many assertions as necessary right now, while developing. As soon as there will be debug builds - and I hope they will be available one day - assertions will become useful.

Every decent FOSS C++ project uses assertions. Why should we go without?

Assertions should be added in all critical places, during development of new code and when fixing errors in order to prevent regressions. During build-upload-test cycles assertions can help to track down errors faster. And when a host-based debugger/simulator is used, assertions help to stop the program at the right place.

algernon commented:
Plugins updated again.

noseglasses commented on src/plugin.h:
```diff
@@ -40,7 +40,7 @@ class Plugin {
     EVENT_CONSUMED,
     ERROR,
-  };
+  }
```
To the best of my knowledge the declaration of an enum class without a semicolon is non standard conforming ([example](https://godbolt.org/g/S16hkY)). I did not try if this works with Arduino, but if so, this is likely to cause problems with future versions of avr-gcc and other compilers.

algernon replied:
Yeah, it did not work. I made the mistake of listening to cpplint, which was telling me that a class does not need a semicolon. Except this is an enum class. A subsequent patch reverted this change.

noseglasses replied:
> which was telling me that a class does not need a semicolon

Which is pretty strange because the standard dictates that a class definition must be followed by a semicolon. This is because of C++ intention to be (mostly) C compatible and there 
```cpp
struct A { } a;
```
was/is legal. The compiler must know if there is a variable definition after the class/struct. Therefore the semicolon.

noseglasses commented on src/plugin.h:
```diff
@@ -40,7 +40,7 @@ class Plugin {
     EVENT_CONSUMED,
     ERROR,
-  }
+  };
```
Sorry, I didn't notice the second commit :wink:

algernon commented:
Some renamings are still WIP, and I need to go over the docs again, once all names are final.

Meanwhile, the [f/plugin-redesign-2.1-squash](https://github.com/keyboardio/Kaleidoscope/tree/f/plugin-redesign-2.1-squash) branch has most of these pulled together, with an improved commit message. Once we're done here, I plan to squash everything into that, perhaps tweak the commit message a bit more, and replace this PR's branch with that one.

noseglasses commented on src/Kaleidoscope.h:
```diff
@@ -147,21 +133,39 @@ class Kaleidoscope_ {
 };
 
-extern Kaleidoscope_ Kaleidoscope;
+extern kaleidoscope::Kaleidoscope_ Kaleidoscope;
```
Class `Kaleidoscope_` is currently ambiguously instanciated both in global namespace and in namespace `kaleidoscope`, [here](https://github.com/keyboardio/Kaleidoscope/blob/af208b0fd35b16531592a08db272acd41a654f7e/src/Kaleidoscope.cpp#L105) and [there](https://github.com/keyboardio/Kaleidoscope/blob/af208b0fd35b16531592a08db272acd41a654f7e/src/Kaleidoscope.cpp#L200).

One of them needs to be removed (preferably the one at global scope).

noseglasses commented on src/Kaleidoscope.h:
```diff
@@ -147,21 +133,39 @@ class Kaleidoscope_ {
+// For compatibility reasons we enable the global variable Kaleidoscope
+// in global namespace.
+//
+extern kaleidoscope::Kaleidoscope_ Kaleidoscope;
```
This could be replace by 
```cpp
extern kaleidoscope::Kaleidoscope_ &Kaleidoscope;
```
and the instanciation [here](https://github.com/keyboardio/Kaleidoscope/blob/af208b0fd35b16531592a08db272acd41a654f7e/src/Kaleidoscope.cpp#L200) by
```cpp
kaleidoscope::Kaleidoscope_ &Kaleidoscope = kaleidoscope::Kaleidoscope;
```
to solve the problem with the redundant instances of class `Kaleidoscope_`, thereby remaining backward compatible.

noseglasses commented on src/kaleidoscope_internal/event_dispatch.h:
```diff
@@ -0,0 +1,141 @@
+  };                                                                          __NL__ \
+                                                                              __NL__ \
+  }                                                                           __NL__ \
+  _REGISTER_EVENT_HANDLER(onSetup,false,())                                   __NL__ \
```
Proposal: Lets use extended [X Macros](https://en.wikipedia.org/wiki/X_Macro) here for greater flexibility with processing hooks in other places (or in the future).

```cpp
// Variadic arguments of _FOR_EACH_EVENT_HANDLER are added to the parameter list of OP.
// This may optionally be used to pass additional information to the operation OP.
//
#define _FOR_EACH_EVENT_HANDLER(OP, ...)                                __NL__ \
   OP(onSetup, false,                                                   __NL__ \
               (),(), ##__VA_ARGS__)                                    __NL__ \
   OP(beforeEachCycle, false,                                           __NL__ \
               (), (), ##__VA_ARGS__)                                   __NL__ \
   OP(onKeyswitchEvent, true,                                           __NL__ \
               (Key &mappedKey, byte row, byte col, uint8_t keyState),  __NL__ \
               (mappedKey, row, col, keyState),                         __NL__ \
                ##__VA_ARGS__)                                          __NL__ \
   OP(beforeReportingState, false,                                      __NL__ \
               (),(),##__VA_ARGS__)                                     __NL__ \
   OP(afterEachCycle, false,                                            __NL__ \
               (),(),##__VA_ARGS__)      

// Invokes _FOR_EACH_EVENT_HANDLER and uses _REGISTER_EVENT_HANDLER for OP
//
_FOR_EACH_EVENT_HANDLER(_REGISTER_EVENT_HANDLER)
```

Note, the `##__VA_ARGS__` ([explanation](https://gcc.gnu.org/onlinedocs/gcc-4.8.2/gcc/Variadic-Macros.html#Variadic-Macros)) which is necessary to be able to pass zero or more additional arguments to `_FOR_EACH_EVENT_HANDLER` that are passed through to the operation `OP`.

Future applications could e.g. be something like this:
```cpp
// This macro has the same signature as _REGISTER_EVENT_HANDLER, except for the additional 
// argument PLUGIN_CLASS. This is generated from an additional variadic argument of 
// _FOR_EACH_EVENT_HANDLER that is passed on to OP where it ends up as PLUGIN_CLASS.
//
#define DEBUG_EVENT_HANDLERS_OF_PLUGIN(HOOK_NAME, SHOULD_ABORT_ON_CONSUMED_EVENT, SIGNATURE, ARGS_LIST, PLUGIN_CLASS) \
   /* Do something special with class PLUGIN_CLASS and its handlers */ \
   ...

_FOR_EACH_EVENT_HANDLER(DEBUG_EVENT_HANDLERS_OF_PLUGIN, kaleidoscope::MyFancyPluginClass)
```

algernon replied:
I think this is roughly what both @obra and me wanted to end up with eventually, but went to sleep before the last mile.

Thanks! :)

noseglasses replied:
As it is way easier for me to write code than to write explanations, I opened a PR #318 that covers the proposed changes.

noseglasses commented:
`src/kaleidoscope/eventhandlerresult.h` might better be renamed `src/kaleidoscope/event_handler_result.h` to be better conforming (and more readable).

algernon commented:
This is no longer WIP, and I force-pushed a squashed together branch.

noseglasses commented:
I know that you are eager to finish this and so am I.

There is one tiny detail about the `_FOR_EACH_EVENT_HANDLER` macro function that I overlooked yesterday.

I added a leading `_` to all the names of macro functions that I considered part of Kaleidoscope's internals (not for use in plugins, etc.). Originally, this included `_FOR_EACH_EVENT_HANDLER` which was initially residing in of one of the headers in `kaleidoscope_internal`. 

After it has been moved to its own header `src/event_handler.h`, it might better be named `FOR_EACH_EVENT_HANDLER`. Thus, marking it as a macro function that's accepted to be used by users, too, for debugging, etc.

obra commented:
Yeah. At least for now, this feels like "not part of the public API"

On Fri, May 11, 2018 at 12:22 AM Gergely Nagy <notifications@github.com>
wrote:

> I'd keep the leading underscore, if for nothing else, to signal that this
> is close to the internals, and should not be used lightly.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope/pull/316#issuecomment-388283018>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACxaPkG9re6dZXI7BQeLTCXksDOyQ1Lks5txTwhgaJpZM4TyrGu>
> .
>

noseglasses commented on src/plugin.h:
```diff
@@ -0,0 +1,103 @@
+#endif
+
+ public:
+  // The following callbacks handle the synchronized communication
```
This text needs to be updated as it is left over from when the event handlers were directly defined in `Plugin`. I think we should refer the reader to `src/event_handlers.h` where all the event handlers are defined and documented.

noseglasses replied:
The the terminology also needs to be updated according to the new glossary.
Also, I wouldn't write about callbacks here, as the term callback is mostly used for stand alone functions and not for class methods, even though they serve the same purpose.

algernon replied:
Nice catch, thanks! Pushed an update that changes this text to a pointer to `event_handlers.h`.

noseglasses commented:
There's still the issue with the [redundant Kaleidoscope instance](https://github.com/keyboardio/Kaleidoscope/pull/316#pullrequestreview-119032244) which is, from my point of view, not urgent but at least worth an issue ticket after V2 has been merged.

As far as I can see class `Kaleidoscope_` does not have any non-static data members right now. So currently the costs for the redundance are zero. Nevertheless, the instance redundancy will hurt in terms of PROGMEM/RAM and also in other ways as soon as there are non static data members added to `Plugin` in future revisions.

algernon commented:
Whoops! Forgot about that. Fixed. Thanks again!

noseglasses commented on src/Kaleidoscope.h:
```diff
@@ -147,21 +133,39 @@ class Kaleidoscope_ {
+// For compatibility reasons we enable the global variable Kaleidoscope
+// in global namespace.
+//
+extern kaleidoscope::Kaleidoscope_ &Kaleidoscope;
```
Not urgent:

Damn! It just occured to me that we could have fixed this with a [using declaration](https://godbolt.org/g/LKKnMZ) to generate a perfect alias and thereby saving 2 bytes of RAM. 
```
using kaleidoscope::Kaleidoscope;
```
If someone changes this, don't forget to remove the instanciation of `&Kaleidoscope` in `kaleidoscope.cpp`.

algernon replied:
Hah! Nice. It also appears to have saved us a few (20) bytes of PROGMEM too.

noseglasses replied:
20 bytes! For a simple reference. I didn't expect that much.

algernon commented:
Back to being WIP:

- We need to deprecate `Kaleidoscope.use()`, the text needs a review.
- We will have plugins that support the V2 API **only**, and should not be used with `Kaleidoscope.use`. These should error out. The `ENFORCE_KALEIDOSCOPE_V2_PLUGIN_API` macro is for this purpose.

We *can* support V1 & V2, but that means I need to update every plugin with wrappers. Mind you, that may end up being the better option... but a v2-only enforcement might still be useful nevertheless.

algernon commented:
Note to self, with regards to the `.use()` deprecation, but applies to `ENFORCE` too:

00:53 **obra** | algernon: something closer to the other use deprecation message, telling them how to do it.
00:53 **obra** | Also possibly add: If this error doesn't make sense to you or you have any trouble, please send a copy of your .ino sketch file to help@keyboard.io and we can help fix it.

algernon commented:
Also, @obra had a sweet idea: have an `UPGRADING.md` file, which details what is deprecated, what is their replacement, and how to migrate. Contains both past and present deprecations, with date of deprecation and date of expected removal.

algernon commented:
- [x] Updated the `Kaleidoscope.use()` deprecation text, reviewed by @obra.
- [x] Made it so that it only fires once per use, and internal uses do not produce a warning.
- [x] Removed `ENFORCE_...`, as I'm going to update all plugins to support v1 too, and that should be the norm during the transition period.
- [ ] `UPGRADE.md` still needs to be written.

algernon commented:
Phew. Should be ready, again.

noseglasses commented:
+1 for the nice `DEPRECATED_USE` macro and its application.

algernon commented:
The text of that is largely @obra's work, all credit to him. And yes, it's <3 <3!

noseglasses commented on doc/glossary.md:
```diff
@@ -0,0 +1,13 @@
+
+Event handler
+
+: A function, usually provided by a `Plugin` that is run by a `Hook`
```
__Plugin__ is not well defined because it is used for the class `Plugin` and also for Arduino libraries, that sometimes don't contain `Plugin` classes. Might be a good candidate for the glossary.
